Merge pull request #123 from jdblischak/error-env-file

jdblischak opened the pull request:
To facilitate debugging similar to my experience in #121 (where my code accidentally deleted the temporary directory and then recreated it), I added an early failure if the environment file is missing.

gaborcsardi commented on R/run.R:
```diff
@@ -1,6 +1,9 @@
 
+  if (!file.exists(env_file))
+    stop("Unable to find environment file in temporary directory. Try restarting R session.")
+
```
I think we should put this in the setup phase somewhere, so that it is also used for the background processes.

jdblischak replied:
@gaborcsardi I moved the check to `default_load_hook()` because that function is called by both `r()` and `r_bg()`. If there is a better place for it, please let me know.

gaborcsardi commented on tests/testthat/test-error.R:
```diff
@@ -242,6 +242,17 @@ test_that("errors in r_session$call() are merged", {
 
+test_that("early error for missing environment file", {
+  tmp <- tempfile()
+  file.rename(env_file, tmp)
```
I think it would be better to run the whole test in a background process, so we don't temporarily mess up the current callr session. I.e. putting the whole thing into a function and callr via `callr::r()`

jdblischak replied:
@gaborcsardi I moved the test to a function and called it with `callr::r()`. Please let me know if the test can be further improved.

jdblischak commented:
@gaborcsardi I updated the PR based on your feedback and added a NEWS bullet. Please review and let me know if there are any further improvements you'd like me to make.

Also, would you like me to rebase and squash my commits onto master and thus remove the merge commit?

gaborcsardi commented:
> Also, would you like me to rebase and squash my commits onto master and thus remove the merge commit?

Yes, please rebase. Thanks!

gaborcsardi commented on R/hook.R:
```diff
@@ -13,6 +13,10 @@ common_hook <- function() {
 default_load_hook <- function(user_hook = NULL) {
+
+  if (!file.exists(env_file))
+    stop("Unable to find environment file in temporary directory. Try restarting R session.")
```
Can you pls break lines be for 75 characters?
```suggestion
    stop(
      "Unable to find environment file in temporary directory.",
      " Try restarting R session."
    )
```

gaborcsardi commented on tests/testthat/test-error.R:
```diff
@@ -242,6 +242,22 @@ test_that("errors in r_session$call() are merged", {
+  f <- function() {
+    tmp <- tempfile()
+    file.rename(callr:::env_file, tmp)
+    on.exit(file.rename(tmp, callr:::env_file))
```
This runs in a throwaway session, so you don't need to restore the file. I suggest just to remove it, that expresses the intent better, I think.

gaborcsardi replied:
Actually, it is fine, no worries.

gaborcsardi commented:
Thanks! Left some very small comments. I agree that the error looks much much better now. 

I wonder if we could do even better, and just put back the env file in this case? That would be robust against `/tmp` cleanup I think.

jdblischak commented:
> Thanks! Left some very small comments. I agree that the error looks much much better now.

I incorporated your latest suggestions, rebased against master, and squashed my commits.

> I wonder if we could do even better, and just put back the env file in this case? That would be robust against /tmp cleanup I think.

I personally prefer the error. If the user's code is somehow deleting the temporary R directory, this is likely a mistake (like it was in my case in #121). The deleted temporary directory could cause later difficult-to-debug problems, so an early failure from callr seems appropriate.

However, if you feel strongly about it, I can update the PR to something like below:

```
  if (!file.exists(env_file)) {
    dir.create(dirname(env_file), recursive = TRUE)
    file.create(env_file)
  }
```

Please let me know what you think.

gaborcsardi commented:
Thanks much!
